src/js/common/zoom.js

```diff
@@ -0,0 +1,71 @@
+{
+
+const filter_in_range_data = (args, range) => {
+  const is_data_in_range = (data, range) => {
+    return data > Math.min(range[0], range[1]) && data < Math.max(range[0], range[1]);
+  }
+
+  return d => {
+    const x_in_range = !('x' in range) || is_data_in_range(d[args.x_accessor], range.x);
+    const y_in_range = !('y' in range) || is_data_in_range(d[args.y_accessor], range.y);
+    return x_in_range && y_in_range;
+  }
+}
+
+const zoom_to_data_domain = (args, range) => {
+  const raw_data = args.raw_data || args.data;
+  if (!('raw_data' in args)) {
+    args.raw_domain = {
+      x: args.scales.X.domain(),
+      y: args.scales.Y.domain(),
+    };
+    args.raw_data = raw_data;
+  }
+  if (args.chart_type === 'point') {
+    if (is_array_of_arrays(raw_data)) {
+      args.data = raw_data.map(function(d) {
+        return d.filter(filter_in_range_data(args, range));
+      });
+    } else {
+      args.data = raw_data.filter(filter_in_range_data(args, range));
+    }
+  }
+  if ('x' in range) args.processed.zoom_x = range.x;
+  else delete args.processed.zoom_x;
+  if ('y' in range) args.processed.zoom_y = range.y;
+  else delete args.processed.zoom_y;
+  new MG.charts[args.chart_type || defaults.chart_type].descriptor(args);
+}
+
+const zoom_to_raw_range = args => {
+  if (!('raw_domain' in args)) return;
+  const range = args.raw_domain;
+  zoom_to_data_domain(args, range);
+  delete args.raw_domain;
+  delete args.raw_data;
+}
+
+const convert_range_to_domain = (args, range) => {
+  const domain = {};
+  if ('x' in range) {
+    const xScale = args.scales.X;
+    domain.x = range.x.map(v => +xScale.invert(v));
+  }
+  if ('y' in range) {
+    const yScale = args.scales.Y;
+    domain.y = range.y.map(v => +yScale.invert(v)).reverse();
+  }
+  return domain;
```

Comment:
Could probably do something similar to what I suggested just above with a forEach loop here.

Reply:
If you need to reduce the number of lines of code this function can be written as follows. And just only avoid one repeat.

```js
const convert_range_to_domain = (args, range) =>
  ['x', 'y'].reduce((domain, dim) => {
    if (!(dim in range)) return domain;
    domain[dim] = range[dim].map(v => +args.scales[dim.toUpperCase()].invert(v));
    if (dim === 'y') domain[dim].reverse();
    return domain;
  }, {});
```
Also it is probably to change
```js
return d => {
  const x_in_range = !('x' in range) || is_data_in_range(d[args.x_accessor], range.x);
  const y_in_range = !('y' in range) || is_data_in_range(d[args.y_accessor], range.y);
  return x_in_range && y_in_range;
}
```
to
```js
  return d => ['x', 'y'].every(dim => !(dim in range) || is_data_in_range(d[args[`${dim}_accessor`]], range[dim]))
```
Also some others.
But does this make the code a bit difficult to understand? :confused: 
I'm not sure if this is necessary?

Reply:
@cnwangjie I think it's pretty understandable, at least to someone who knows what they're looking for. In general I prefer this kind of "denser" style of code-- it requires a bit more functional programming knowledge to understand, but it's much more concise.

If you think a particular section is unclear, it never hurts to put in a comment like this:

```js
// Converts a selection range into a set of coordinates that we can use to zoom
// the chart to a particular region
const convert_range_to_domain = (args, range) =>
  ...
```